### ISSUE TITLE
Prevent loading all image meta data

### ISSUE THREAD

**Issue by webzunft** [image-source-control/image-source-control#117]
In https://github.com/image-source-control/image-source-control/blob/336e5a20ab77af958253dae0941966d638a52ba3/includes/gutenberg/gutenberg.php#L36, ISC loads the meta data for all images into a JS array. This could cause problems with many images.

I cannot remember why we did this, but my feeling would be that we did it to have the data available dynamically when an image source is added. Still, we might want to change that into using an AJAX request per image.

**Comment by webzunft** [comment]
This [thread on wp.org](https://wordpress.org/support/topic/critical-error-when-visiting-widgets-settings-in-admin/) confirms my fear that loading the meta data can cause a timeout with many images.

@cnhk can you remember why you implemented the loop like this and do you have an idea about how to resolve it?

**Comment by cnhk** [comment]
> can you remember why you implemented the loop like this and do you have an idea about how to resolve it?

This is the fastest way, having all the data ready when an image is clicked on. But I didn't expect that to consume 256MB of memory. The only alternative is to load data when they are needed. But that would be really slow for a site with only few images.

May be  we should add a setting to enable/disable that, or just fix a limit on image count above which, data will be fetched on demand.

**Comment by webzunft** [comment]
Is there already a mechanism to fetch the data individually?

**Comment by cnhk** [comment]
There are a new set of functions to read and write meta with the API v2, but I never tried them.

**Comment by webzunft** [comment]
@cnhk could you write a quick solution that would only query meta data from ISC? In the use case above I believe they don’t use ISC yet, but since the query looks for any meta data, it will return a lot of results.

**Comment by cnhk** [comment]
@webzunft https://github.com/image-source-control/image-source-control/pull/158

This is a "quick" solution. It seems that using the new functions requires to change the block attributes (could be quite complex given that there are already people using the block, and the lack of documentation).

In short:
 - once batch call for all image detected at page load. I didn't found an "appropriate" JS hook so it's technically a timeout. ISC fields for Images/cover/gallery blocks that are loaded successively with less than one second of interval (there is more than enough margin here) are loaded in batch in one single AJAX call.
 - blocks that are loaded after that (like inserting a new image/gallery) will use one separate call per image.

**Comment by webzunft** [comment]
This fix originated from [this issue](https://wordpress.org/support/topic/critical-error-when-visiting-widgets-settings-in-admin/) reported on wp.org. According to the user, it is still not gone.
@cnhk do you see a query in your new or in the existing code that could cause this timeout?

**Comment by cnhk** [comment]
Out of memory in `wp-db.php`, this is still the postmeta. Can we get an estimate on how many images are on the widgets page? All we can do at this point is to load the post meta by packet. But we need to determine how much we can load per ajax call.

**Comment by cnhk** [comment]
> But we need to determine how much we can load per ajax call.

Or we can just load incrementally until we are nearly at the limit in memory|execution time,

**Comment by webzunft** [comment]
They just use 4 images. How could we learn more here? I guess Query Monitor wouldn’t help either.

**Comment by cnhk** [comment]
Can you get the line where the memory error occurs like the last time? It still looks like a lots of images are loaded.

**Comment by Galanos** [comment]
Remember that he offered to ”give us a tour“, maybe it's time to take that offer.

**Comment by cnhk** [comment]
That would be the fastest way

**Comment by webzunft** [comment]
> Can you get the line where the memory error occurs like the last time?

As I see it, the error changed from pointing to a specific line in ISC to a general DB timeout. The user also posted an update on his setup. He even – very kind of him – redid the setup with the same result.

**Comment by cnhk** [comment]
There is still something loading the postmetas on the widgets page, but it's not from the fields we added to image/cover/gallery blocks.

I set up a media library with 1000+ images. The query when counting missing sources starts to be quite heavy. But there is no problem on the widgets page with just query monitor and ISC. I tried with the 1k+ images not and then on the front end. No effect.

It might be a setting/plugin on his installation. So at this point, I think we need an access to the site.

**Comment by webzunft** [comment]
The problem is that we can’t ask for access to a site when people come through the forum. That is strictly against the terms on wordpress.org.
The user wrote that using the Classic Widget plugin "resolves" the issue, so I guess it is not about the query that looks for missing sources.
The user is comfortable testing a new version also from GitHub. Any ideas about what we can check?

**Comment by cnhk** [comment]
> The user wrote that using the Classic Widget plugin "resolves" the issue, so I guess it is not about the query that looks for missing sources.

That means it about with Gutenberg.

Can you ask him to install query monitor to see if there is any error?

There is likely another plugin that is modifying images blocks like we do with ISC. And since he proposed himself the dashboard access 

![image](https://user-images.githubusercontent.com/3407174/162426156-6b372ec8-5bc6-4e39-ad01-b4ceba6ff818.png)

we might need this if there is nothing obvious in query monitor

**Comment by Galanos** [comment]
This might be unrelated, but:
I had wondered if maybe our **admin notice about missing sources** might be the culprit (although this wouldn’t explain why it only appears on the widget page), so I retraced the functions back to …
- [get_attachments_with_empty_sources()](https://github.com/image-source-control/image-source-control/blob/ffd23947d8db7552d48188c7097a87679818f4b8/includes/model.php#L372) and
- [get_unused_attachments()](https://github.com/image-source-control/image-source-control/blob/ffd23947d8db7552d48188c7097a87679818f4b8/includes/model.php#L404)

These functions seem to pull the **full datasets** (of all of 64k images that have ISC metadata attached to them) from the posts table and then in …
- [count_missing_sources()](https://github.com/image-source-control/image-source-control/blob/ffd23947d8db7552d48188c7097a87679818f4b8/includes/model.php#L355)
… the returned arrays get counted by PHP’s count() function. That’s thousands upon thousands of database elements retrieved to just count the number.

Isn’t there a way to use [MySQL’s COUNT() function](https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions.html#function_count) here? That would be waaay better for performance. Or doesn’t WordPress facilitate this?

Please tell me if I’ve overlooked anything.

**Comment by webzunft** [comment]
@Galanos shouldn’t be related, but worth a new issue.
@cnhk since the user agreed to run a test version, how about we narrow it down by disabling the ISC meta data loading in Gutenberg and ask the user to test it. Of course, it will be broken, but then we at least know that it is related to ISC loading meta data.
We cannot get a login from the user. It is not even allowed to offer one on wp.org. The user is violating wp.org terms.

**Comment by cnhk** [comment]
@webzunft Here is the version without calling the REST API:

https://github.com/cnhk/image-source-control/tree/dev-no-post-meta

I also noticed a weird error with the gallery block that only happens with ISC enabled (on all pages, not only the widgets so probably not related, also maybe because I have 1000+ images on the library). Have any one faced a problem with the gallery recently? I tried to sort it out using `SCRIPT_DEBUG` but the log is 20+ kb long. So if someone have some info easier to understand I'll gladly take that.

**Comment by Galanos** [comment]
(Somehow my e-mail reply didn’t go through …)
@cnhk What does the weird error look like? I haven’t noticed anything but I’d like to try and replicate it.

**Comment by cnhk** [comment]
@Galanos When you have a gallery block in a page, and click on one image to edit one of the field in the sidebar, the block crashes and you get the "This block has encountered an error and cannot be previewed." on all the images.

But I tried to get some screenshot today, but it didn't happened anymore. I tried with both the public version, the current `master` branch, and the one that doesn't load ISC field I mentioned in my previous message with no luck. I also tried with installation with only few images then 500+ images in the library. Still got nothing.

**Comment by Galanos** [comment]
@webzunft @cnhk The user answered: the most recent test version that does **not** call the REST API does **not** cause the error to appear.

@cnhk Sorry, last week I didn’t get to looking into the weird gallery problem, and then we had a long Easter weekend. I’ll try to look into this early this week.

**Comment by cnhk** [comment]
There is a block that is loading all the ISC fields for all the images on that widgets page. Can we get what blocks exactly (from which plugins or theme) are in use on the widgets page?

**Comment by Galanos** [comment]
@cnhk I just wrote to the user and asked him to use the “[Health Check & Troubleshooting](https://wordpress.org/plugins/health-check/)” plugin to test his site with _only_ ISC active, no other plugin. I’d recommend we wait for his answer.

**Comment by cnhk** [comment]
It needs to be tested with the current master branch in that case (the one that load ISC fields using WordPress REST API), not the dev version nor the one from wordpress.org

**Comment by Galanos** [comment]
@cnhk That should be the one that he deactivated in order to try out the latest one but I wrote a short follow-up to make sure. Thanks!

**Comment by Galanos** [comment]
@cnhk Concerning the weird gallery bug—I installed three different versions of ISC …
- public 2.6.0
- current `master` branch
- custom version without REST API
… but the error you described did not appear. My installation of WP had ~100 images in the media library and 6 images in the gallery block.

**Comment by Galanos** [comment]
@webzunft @cnhk The user answered. He installed the Health Check plugin and checked: the error still occurs.

> I performed the actions with the Health Check plugin but I still get the error. To identify the plugin which caused this I also tested by deactivating plugins and only have Image Source Control active, that's how I found out that ISC caused the issue. 

You got a next step in mind? Should I ask about the [blocks from themes](https://github.com/image-source-control/image-source-control/issues/117#issuecomment-1102504711) (plugins were deactivated) that are active on the widget page?

**Comment by webzunft** [comment]
As discussed personally with @Galanos:

- we cannot investigate deeper without reproducing the issue on our end or having access to a staging environment which we can break
- #170 could be a temporary solution, if that helps the user

**Comment by cnhk** [comment]
> Should I ask about the blocks from themes (plugins were deactivated) that are active on the widget page?

A copy of the theme would be ideal if possible. Since all plugins were disabled, there is a great chance that the theme is loading these meta data.

> You got a next step in mind? 

I thinks we can adjust a bit the data loading using the rest API. Rather than loading the fields for all the images on the post or widget pages, we can spit the call into multiple sub calls when it's needed. But let's wait for the theme first.

**Comment by Galanos** [comment]
Collecting some info from other issues, PRs, and Slack here:
- **in parallel** events, this issue’s fix (#158)—which switched the loading of ISC image metadata in the block editor first to AJAX, then to the REST API—caused issue #171 which was fixed by #173
- **next in this issue**’s events, we received access to a staging site displaying the issue of this thread
- **there**, we were able to witness the error on the Widgets page
- **there**, we installed branch #173, which worked fine—the error disappeared
- **there**, further checks by the user and by us (manual tests and _Query Monitor_ results) turned up nothing special
- **next in this issue**’s events, up to this point, all changes have been merged into the master branch
- **coming up next**, we need to address #172 before releasing

**Comment by Galanos** [comment]
Final update before closing:
- #172 has been solved
- all changes that have been relevant to this issue have been committed, merged, and released in 2.7.0
- no error reports have reached us since then, 9 days ago

**Comment by Galanos** [comment]
@cnhk Could you describe the bug in the gallery? Didn't notice anything but I'd like to replicate it.

**Comment by cnhk** [comment]
@Galanos It's like I described in my [previous comment](https://github.com/image-source-control/image-source-control/issues/117#issuecomment-1096315589). I saw that one day, but I've never been able to reproduce it.

The _"This block has encountered an error and cannot be previewed."_ error sometime means the data (block attributes) stored in the post content doesn't match the attributes described in the block's JS. So I suppose it was be cause of something I tested/debugged while working on this issue.
